### DISCUSSION TITLE
func.distinct behaves differently with Multiple Columns

### DISCUSSION THREAD

**Discussion by simonnelse** [sqlalchemy/sqlalchemy#9481] [Usage Questions] [answered]
`func.distinct` function in SQLAlchemy is behaving differently when used with multiple columns.
### PostgreSQL query:
```sql
SELECT * FROM foo;
 col1 | col2 
------+------
    1 | A
    2 | B
    1 | A
(3 rows)

SELECT array_agg(distinct(foo.col1)) AS result_1 FROM foo;
 result_1 
-------------
 {1,2}

SELECT array_agg(distinct(foo.col1, foo.col2)) AS result_2 FROM foo;
 result_2 
------------
{"(1,A)","(2,B)"}
```
The above output is as expected, but while using SQLAlchemy for the same goal, the result is as follows:
### SQLAlchemy output:
```sql
stmt_1 = select(
    func.array_agg(func.distinct(foo.c.col1))
) 
result_1 = engine.execute(stmt_1).fetchall()
# result_1 is [([1, 2],)]

stmt_2 = select(
    func.array_agg(func.distinct(foo.c.col1, foo.c.col2))
)
result_2 = engine.execute(stmt_2).fetchall()
# result_2 is [(['{', '"', '(', '1', ',', 'A', ')', '"', ',', '"', '(', '2', ',', 'B', ')', '"', '}'],)]
```

Can you help me understand why the the **func.distinct** function in SQLAlchemy acts differently with multiple columns compared to PostgreSQL? I'm puzzled as to why the result of **result_2** isn't the same as the PostgreSQL output, and wondering if there's anything I might have missed. Thanks a bunch!

_Using SQLAlchemy V1.4.34_

**Comment by zzzeek** [comment, marked as answer]
these nested arrays are hard, the drivers dont know what to do with them by default.    This looks like it becomes an ARRAY of Tuple, which the drivers deliver as a mess.  run it with NullType to see

```py
    from sqlalchemy.types import NullType

    stmt_2 = select(
        func.array_agg(func.distinct(A.data1, A.data2), type_=NullType)
    )
```

this is the raw row:

```
 ('{"(asfd,fggs)"}',)
```


SQLAlchemy's ARRAY type doesn't know how to work with that.   There would have to be some CAST or other application on the server side so that the driver gets back something it can parse into the format that SQLAlchemy expects.

I tried

```
    stmt_2 = select(
        cast(
            func.array_agg(func.distinct(A.data1, A.data2)),
            ARRAY(String, dimensions=1)
        )

    )

```

~~but this doesn't work either~~.  I think the above works for what you are requesting, see next comment.

**Reply by zzzeek** [reply]
mm i think the second idea works, here is a compairson:

```py
from sqlalchemy.dialects.postgresql import array

with e.connect() as conn:
    stmt_1 = select(
        array(["(abc, def)"])
    )
    result_1 = conn.execute(stmt_1).fetchall()

    stmt_2 = select(
        cast(
            func.array_agg(func.distinct(A.data1, A.data2)),
            ARRAY(String, dimensions=1)
        )

    )
    result_2 = conn.execute(stmt_2).fetchall()
    print(result_2[0][0])

```

output:

```
['(abc, def)']
['(asfd,fggs)']
```

so you want the tuple as a string, that's it

**Comment by CaselIT** [comment]
This is very depended on the dialect in use.

psycog3 accepts the version with `sa.func.array_agg(sa.func.distinct(sa.tuple_(...)))` while psycopg2 does not.
At work I use this using array and I have no problem at least with psycopg2
```
from sqlalchemy.dialects.postgresql import array
sa.func.array_agg(sa.func.distinct(array([...])))
```

it seems to be working also with psycopg3
